File: app/src/main/java/djordjeh/architecture/mvp/ui/BasePresenterImpl.java
package djordjeh.architecture.mvp.ui;

import android.support.annotation.NonNull;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

abstract public class BasePresenterImpl<V> implements BasePresenter {

    @NonNull
    protected final V view;

    @NonNull
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public BasePresenterImpl(@NonNull V view) {
        this.view = view;
    }

    @Override
    public void stop() {
        compositeDisposable.clear();
    }

    @Override
    public void destroy() {
        if (!this.compositeDisposable.isDisposed()) {
            this.compositeDisposable.dispose();
        }
    }

    protected void addDisposable(Disposable disposable) {
        if (!this.compositeDisposable.isDisposed()) {
            this.compositeDisposable.add(disposable);
        }
    }

    protected boolean removeDisposable(Disposable disposable) {
        return this.compositeDisposable.remove(disposable);
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/ui/BasePresenterImpl.kt
package djordjeh.architecture.mvp.ui

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BasePresenterImpl<V>(protected val view: V) : BasePresenter {

    private val compositeDisposable = CompositeDisposable()

    override fun stop() {
        compositeDisposable.clear()
    }

    override fun destroy() {
        if (!this.compositeDisposable.isDisposed) {
            this.compositeDisposable.dispose()
        }
    }

    protected fun addDisposable(disposable: Disposable) {
        if (!this.compositeDisposable.isDisposed) {
            this.compositeDisposable.add(disposable)
        }
    }

    protected fun removeDisposable(disposable: Disposable): Boolean {
        return this.compositeDisposable.remove(disposable)
    }
}
